gon/contour.py

```diff
@@ -0,0 +1,148 @@
+from fractions import Fraction
+from functools import reduce
+from typing import (Iterable,
+                    List,
+                    Sequence)
+from weakref import WeakKeyDictionary
+
+from bentley_ottmann.planar import edges_intersect
+from memoir import cached
+from reprit.base import generate_repr
+from robust.hints import Expansion
+from robust.utils import (sum_expansions,
+                          two_product,
+                          two_two_diff)
+
+from gon.angular import (Orientation,
+                         to_orientation)
+from gon.geometry import Geometry
+from gon.hints import Coordinate
+from gon.point import (Point,
+                       RawPoint)
+
+MIN_VERTICES_COUNT = 3
+
+Vertices = Sequence[Point]
+RawContour = List[RawPoint]
+
+
+class Contour(Geometry):
+    __slots__ = '_vertices',
+
+    def __init__(self, vertices: Iterable[Point]) -> None:
+        self._vertices = tuple(vertices)
+
+    __repr__ = generate_repr(__init__)
+
+    def __hash__(self) -> int:
+        return hash(self._vertices)
+
+    def __eq__(self, other: 'Geometry') -> int:
+        return (self._vertices == other._vertices
+                if isinstance(other, Contour)
+                else NotImplemented)
+
+    @property
+    def vertices(self) -> Vertices:
+        return self._vertices
+
+    @cached.map_(WeakKeyDictionary())
+    def raw(self) -> RawContour:
+        return [vertex.raw() for vertex in self._vertices]
+
+    @classmethod
+    def from_raw(cls, raw: RawContour) -> 'Contour':
+        return cls(Point.from_raw(raw_vertex) for raw_vertex in raw)
+
+    @cached.property_
+    def normalized(self) -> 'Contour':
+        vertices = self._vertices
+        min_index = min(range(len(vertices)),
+                        key=vertices.__getitem__)
+        return Contour(vertices[min_index:] + vertices[:min_index])
+
+    @cached.property_
+    def orientation(self) -> 'Orientation':
+        vertices = self.normalized.vertices
+        return to_orientation(vertices[0], vertices[-1], vertices[1])
+
+    def to_clockwise(self) -> 'Contour':
+        return (self
+                if self.orientation is Orientation.CLOCKWISE
+                else self.reverse())
+
+    def to_counterclockwise(self) -> 'Contour':
+        return (self
+                if self.orientation is Orientation.COUNTERCLOCKWISE
+                else self.reverse())
+
+    def reverse(self) -> 'Contour':
+        vertices = self._vertices
+        return Contour(vertices[:1] + vertices[:0:-1])
```

Comment:
How about leaving only `reverse` and making a `is_ccw` method to check if it's oriented counterclockwise?

Reply:
in my particular use-case I need to give contour a specific orientation, making checks can be done on the upper level though, not so sure about `is_ccw` method/property, do we need `is_cw` then too? but they are can be boiled down to `contour.orientation is Orientation.CLOCKWISE`, it may be to many words at first, but looks perfectly readable to me, we can add those methods/properties afterwards